File: sdk/typescript/src/billing.ts
import { createClient, type Client, type Transport } from "@connectrpc/connect";
import { create } from "@bufbuild/protobuf";
import { BillingCommandController } from "@stigmer/protos/ai/stigmer/billing/v1/command_pb";
import { BillingQueryController } from "@stigmer/protos/ai/stigmer/billing/v1/query_pb";
import {
  GetOrCreateBillingAccountInputSchema,
  GetBillingAccountInputSchema,
  GetCreditBalanceInputSchema,
  AdjustCreditsInputSchema,
  GetCreditLedgerInputSchema,
  CreateCreditCheckoutSessionInputSchema,
  CreateBillingPortalSessionInputSchema,
  SetAutoRechargeConfigInputSchema,
  GetBillingUsageReportInputSchema,
  GetCustomerModelPricingInputSchema,
  GetModelPricingGovernanceInputSchema,
  DecideModelPricingOverrideInputSchema,
  UpsertModelPricingBaselineInputSchema,
  RetireModelPricingBaselineInputSchema,
  ListModelPricingBaselinesInputSchema,
  type CreateCreditCheckoutSessionResponse,
  type CreateBillingPortalSessionResponse,
  type CreditLedgerResponse,
  type BillingUsageReportResponse,
  type CustomerModelPricingResponse,
  type ModelPricingGovernanceResponse,
  type ModelPricingBaselinesResponse,
} from "@stigmer/protos/ai/stigmer/billing/v1/io_pb";
import type { CreditLedgerEntry } from "@stigmer/protos/ai/stigmer/billing/v1/credit_pb";
import type { ModelPricingOverride } from "@stigmer/protos/ai/stigmer/billing/v1/pricing_override_pb";
import type { ModelPricingBaseline } from "@stigmer/protos/ai/stigmer/billing/v1/model_pricing_baseline_pb";
import type { BillingAccount, CreditBalance } from "@stigmer/protos/ai/stigmer/billing/v1/billing_account_pb";
import type { LedgerEntryType, LedgerView } from "@stigmer/protos/ai/stigmer/billing/v1/enum_pb";
import { PageInfoSchema } from "@stigmer/protos/ai/stigmer/commons/rpc/pagination_pb";
import { timestampFromDate } from "@bufbuild/protobuf/wkt";
import { wrapError } from "./gen/errors.js";

/** Parameters for creating a Stripe Checkout Session. */
export interface CreateCheckoutSessionParams {
  readonly orgId: string;
  readonly packId: string;
  readonly successUrl: string;
  readonly cancelUrl: string;
}

/** Parameters for creating a Stripe Billing Portal session. */
export interface CreateBillingPortalSessionParams {
  readonly orgId: string;
  readonly returnUrl: string;
}

/** Parameters for configuring auto-recharge. */
export interface SetAutoRechargeConfigParams {
  readonly orgId: string;
  readonly enabled: boolean;
  readonly thresholdMicros: bigint;
  readonly rechargeAmountMicros: bigint;
  readonly monthlyCapMicros: bigint;
}

/** Parameters for a manual credit adjustment. */
export interface AdjustCreditsParams {
  readonly orgId: string;
  /** Positive to add credits, negative to remove. */
  readonly amountMicros: bigint;
  /** Human-readable reason recorded on the ledger entry (audit trail). */
  readonly reason: string;
  /** Client-supplied deduplication key to prevent double-processing. */
  readonly idempotencyKey: string;
}

/** Parameters for querying the credit ledger. */
export interface GetCreditLedgerParams {
  readonly orgId: string;
  /** Pagination: `{ num, size }` where `num` is the 0-based page number. */
  readonly page?: { readonly num: number; readonly size: number };
  readonly typeFilter?: LedgerEntryType[];
  /** Filter to entries on or after this timestamp. */
  readonly startTime?: Date;
  /** Filter to entries on or before this timestamp. */
  readonly endTime?: Date;
  /**
   * Server-resolved ledger slice. When set to `LedgerView.statement`, the
   * server returns only customer-facing money-movement entry types and
   * excludes internal mechanics (per-call usage debits, reservation
   * holds/releases). Defaults to the full ledger.
   */
  readonly view?: LedgerView;
}

/** Parameters for querying the billing usage report. */
export interface GetBillingUsageReportParams {
  readonly orgId: string;
  readonly startTime: Date;
  readonly endTime: Date;
}

/** Parameters for querying customer model pricing. */
export interface GetCustomerModelPricingParams {
  readonly orgId?: string;
}

/** Parameters for creating or revising a model registry baseline entry. */
export interface UpsertModelPricingBaselineParams {
  /**
   * The baseline entry to create or revise, keyed by
   * (modelId, provider, harness). Lifecycle fields (baselineId, status,
   * decision stamps, pricing effectiveAt) are server-owned and ignored.
   */
  readonly baseline: ModelPricingBaseline;
  /** Optional operator note recorded on the revision for the audit trail. */
  readonly revisionNote?: string;
}

/** Parameters for retiring a model from the registry catalog. */
export interface RetireModelPricingBaselineParams {
  readonly modelId: string;
  readonly provider: string;
  readonly harness: string;
  /** Optional operator note recorded on the retirement. */
  readonly revisionNote?: string;
}

/** Parameters for listing the model registry baseline catalog. */
export interface ListModelPricingBaselinesParams {
  /**
   * When `true`, includes SUPERSEDED and RETIRED revisions (the full
   * audit history). Default: ACTIVE documents only.
   */
  readonly includeHistory?: boolean;
}

/** Parameters for deciding a pending pricing override. */
export interface DecideModelPricingOverrideParams {
  /** The PENDING_SIGNOFF override to decide. */
  readonly overrideId: string;
  /**
   * `true` approves (the override becomes ACTIVE and supersedes any
   * current ACTIVE override on the same pricing key); `false` rejects.
   */
  readonly approve: boolean;
  /** Optional note recorded on the decision for the audit trail. */
  readonly decisionNote?: string;
}

/**
 * Client for the billing bounded context.
 *
 * Wraps the user-facing billing RPCs: account provisioning,
 * balance queries, ledger history, and credit purchases via
 * Stripe Checkout. Internal execution-billing RPCs (authorize,
 * report, finalize) are not exposed — they are called only by
 * the Temporal workflow and agent runner.
 */
export class BillingClient {
  private readonly command: Client<typeof BillingCommandController>;
  private readonly query: Client<typeof BillingQueryController>;

  constructor(transport: Transport) {
    this.command = createClient(BillingCommandController, transport);
    this.query = createClient(BillingQueryController, transport);
  }

  /**
   * Provision or retrieve the billing account for an organization.
   *
   * Idempotent: creates the account on first call, returns the
   * existing account on subsequent calls.
   */
  async getOrCreateBillingAccount(orgId: string): Promise<BillingAccount> {
    try {
      return await this.command.getOrCreateBillingAccount(
        create(GetOrCreateBillingAccountInputSchema, { orgId }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /** Retrieve the billing account for an organization. */
  async getBillingAccount(orgId: string): Promise<BillingAccount> {
    try {
      return await this.query.getBillingAccount(
        create(GetBillingAccountInputSchema, { orgId }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /** Retrieve the credit balance breakdown for an organization. */
  async getCreditBalance(orgId: string): Promise<CreditBalance> {
    try {
      return await this.query.getCreditBalance(
        create(GetCreditBalanceInputSchema, { orgId }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Manually adjust an organization's credit balance.
   *
   * Positive `amountMicros` adds credits (e.g. funding a tenant org),
   * negative removes them. The adjustment is recorded as a ledger entry
   * with the supplied reason; the idempotency key deduplicates retries.
   * Requires `can_manage_billing` on the org.
   */
  async adjustCredits(params: AdjustCreditsParams): Promise<CreditLedgerEntry> {
    try {
      return await this.command.adjustCredits(
        create(AdjustCreditsInputSchema, {
          orgId: params.orgId,
          amountMicros: params.amountMicros,
          reason: params.reason,
          idempotencyKey: params.idempotencyKey,
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /** Retrieve paginated credit ledger entries with optional filters. */
  async getCreditLedger(params: GetCreditLedgerParams): Promise<CreditLedgerResponse> {
    try {
      return await this.query.getCreditLedger(
        create(GetCreditLedgerInputSchema, {
          orgId: params.orgId,
          ...(params.page && {
            page: create(PageInfoSchema, {
              num: params.page.num,
              size: params.page.size,
            }),
          }),
          ...(params.typeFilter?.length && { typeFilter: params.typeFilter }),
          ...(params.startTime && { startTime: timestampFromDate(params.startTime) }),
          ...(params.endTime && { endTime: timestampFromDate(params.endTime) }),
          ...(params.view !== undefined && { view: params.view }),
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Create a Stripe Checkout Session to purchase a credit pack.
   *
   * Returns the Stripe-hosted checkout URL. The caller should
   * redirect the user to `checkoutUrl` to complete payment.
   * Credits are provisioned asynchronously via webhook after
   * payment succeeds.
   */
  async createCreditCheckoutSession(
    params: CreateCheckoutSessionParams,
  ): Promise<CreateCreditCheckoutSessionResponse> {
    try {
      return await this.command.createCreditCheckoutSession(
        create(CreateCreditCheckoutSessionInputSchema, {
          orgId: params.orgId,
          packId: params.packId,
          successUrl: params.successUrl,
          cancelUrl: params.cancelUrl,
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Create a Stripe Billing Portal session for payment method management.
   *
   * Returns the Stripe-hosted portal URL. The caller should redirect
   * the user to `portalUrl` to manage their saved payment methods.
   * Changes made in the portal are synced back via webhooks.
   */
  async createBillingPortalSession(
    params: CreateBillingPortalSessionParams,
  ): Promise<CreateBillingPortalSessionResponse> {
    try {
      return await this.command.createBillingPortalSession(
        create(CreateBillingPortalSessionInputSchema, {
          orgId: params.orgId,
          returnUrl: params.returnUrl,
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Configure automatic credit recharge for an organization.
   *
   * When enabled, the billing system charges the account's saved
   * payment method whenever the available balance drops below the
   * configured threshold. Returns the updated BillingAccount.
   */
  async setAutoRechargeConfig(
    params: SetAutoRechargeConfigParams,
  ): Promise<BillingAccount> {
    try {
      return await this.command.setAutoRechargeConfig(
        create(SetAutoRechargeConfigInputSchema, {
          orgId: params.orgId,
          enabled: params.enabled,
          thresholdMicros: params.thresholdMicros,
          rechargeAmountMicros: params.rechargeAmountMicros,
          monthlyCapMicros: params.monthlyCapMicros,
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Retrieve an aggregated billing usage report for a date range.
   *
   * Returns total provider cost, total billable amount, execution
   * and LLM call counts, and a per-model breakdown with cost tier
   * attribution. Data is sourced from the `llm_call_usage_record`
   * collection (proxy-observed, tamper-proof).
   */
  async getBillingUsageReport(
    params: GetBillingUsageReportParams,
  ): Promise<BillingUsageReportResponse> {
    try {
      return await this.query.getBillingUsageReport(
        create(GetBillingUsageReportInputSchema, {
          orgId: params.orgId,
          startTime: timestampFromDate(params.startTime),
          endTime: timestampFromDate(params.endTime),
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Retrieve the customer-facing model price list with markup applied.
   *
   * Returns per-million-token prices for all billable models, with
   * the active billing policy markup already factored in. These are
   * the prices the customer pays, organized by harness and cost tier.
   *
   * Pass `orgId` to resolve org-specific policy overrides (future).
   * Omit for default pricing.
   */
  async getCustomerModelPricing(
    params?: GetCustomerModelPricingParams,
  ): Promise<CustomerModelPricingResponse> {
    try {
      return await this.query.getCustomerModelPricing(
        create(GetCustomerModelPricingInputSchema, {
          orgId: params?.orgId ?? "",
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Retrieve the platform pricing governance view: baseline vs effective
   * rates per model, ACTIVE override provenance, and pending sign-off
   * proposals from the pricing feedback loop.
   *
   * Platform-operator surface (`can_manage_model_pricing` on
   * `platform:stigmer`): rates are raw provider prices, pre-markup.
   */
  async getModelPricingGovernance(): Promise<ModelPricingGovernanceResponse> {
    try {
      return await this.query.getModelPricingGovernance(
        create(GetModelPricingGovernanceInputSchema, {}),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Record a human decision on a PENDING_SIGNOFF pricing override.
   *
   * Approving makes the override ACTIVE (superseding any current ACTIVE
   * override on the same pricing key) and recomposes the effective
   * registry; rejecting archives it for audit. Returns the decided
   * override with the decision stamped.
   */
  async decideModelPricingOverride(
    params: DecideModelPricingOverrideParams,
  ): Promise<ModelPricingOverride> {
    try {
      return await this.command.decideModelPricingOverride(
        create(DecideModelPricingOverrideInputSchema, {
          overrideId: params.overrideId,
          approve: params.approve,
          decisionNote: params.decisionNote ?? "",
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Retrieve the model registry baseline catalog: ACTIVE entries by
   * default, or the full append-only revision history with
   * `includeHistory`.
   *
   * Platform-operator surface (`can_manage_model_pricing` on
   * `platform:stigmer`): rates are raw provider prices, pre-markup.
   */
  async listModelPricingBaselines(
    params?: ListModelPricingBaselinesParams,
  ): Promise<ModelPricingBaselinesResponse> {
    try {
      return await this.query.listModelPricingBaselines(
        create(ListModelPricingBaselinesInputSchema, {
          includeHistory: params?.includeHistory ?? false,
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Create or revise one model registry baseline entry (catalog + list
   * prices). Append-only: an existing ACTIVE entry for the same
   * (modelId, provider, harness) key is superseded, never mutated, and
   * the effective registry recomposes immediately. Returns the new
   * revision with server-stamped lifecycle fields.
   */
  async upsertModelPricingBaseline(
    params: UpsertModelPricingBaselineParams,
  ): Promise<ModelPricingBaseline> {
    try {
      return await this.command.upsertModelPricingBaseline(
        create(UpsertModelPricingBaselineInputSchema, {
          baseline: params.baseline,
          revisionNote: params.revisionNote ?? "",
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }

  /**
   * Retire one model from the registry catalog. The model disappears
   * from every price surface on the next composition pass; the document
   * is kept for audit and the key can be revived by a subsequent upsert.
   */
  async retireModelPricingBaseline(
    params: RetireModelPricingBaselineParams,
  ): Promise<ModelPricingBaseline> {
    try {
      return await this.command.retireModelPricingBaseline(
        create(RetireModelPricingBaselineInputSchema, {
          modelId: params.modelId,
          provider: params.provider,
          harness: params.harness,
          revisionNote: params.revisionNote ?? "",
        }),
      );
    } catch (e) {
      throw wrapError(e);
    }
  }
}

File: sdk/go/billing.go
package stigmer

import (
	"context"
	"time"

	"github.com/stigmer/stigmer/sdk/go/v3/internal/gen"
	billingv1 "github.com/stigmer/stigmer/sdk/go/v3/proto/ai/stigmer/billing/v1"
	rpc "github.com/stigmer/stigmer/sdk/go/v3/proto/ai/stigmer/commons/rpc"
	"google.golang.org/grpc"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// Proto payload re-exports for the billing bounded context. Response
// messages are returned as-is (the search.go convention): the proto types
// are the contract, and hand-mapped mirrors would only drift from it.
type (
	// BillingAccount is an organization's billing account with balance and auto-recharge config.
	BillingAccount = billingv1.BillingAccount
	// CreditBalance is the credit balance breakdown (available, reserved, total).
	CreditBalance = billingv1.CreditBalance
	// CreditLedgerEntry is one immutable credit ledger entry.
	CreditLedgerEntry = billingv1.CreditLedgerEntry
	// CreditLedgerResponse is a paginated page of ledger entries.
	CreditLedgerResponse = billingv1.CreditLedgerResponse
	// BillingUsageReportResponse is an aggregated usage report for a date range.
	BillingUsageReportResponse = billingv1.BillingUsageReportResponse
	// CreateCreditCheckoutSessionResponse holds the Stripe-hosted checkout URL.
	CreateCreditCheckoutSessionResponse = billingv1.CreateCreditCheckoutSessionResponse
	// CreateBillingPortalSessionResponse holds the Stripe-hosted portal URL.
	CreateBillingPortalSessionResponse = billingv1.CreateBillingPortalSessionResponse
	// CustomerModelPricingResponse is the customer-facing model price list.
	CustomerModelPricingResponse = billingv1.CustomerModelPricingResponse
	// ModelPricingGovernanceResponse is the operator pricing governance view.
	ModelPricingGovernanceResponse = billingv1.ModelPricingGovernanceResponse
	// ModelPricingBaselinesResponse is the model registry baseline catalog.
	ModelPricingBaselinesResponse = billingv1.ModelPricingBaselinesResponse
	// ModelPricingBaseline is one model registry baseline entry (catalog + list prices).
	ModelPricingBaseline = billingv1.ModelPricingBaseline
	// ModelPricingOverride is one pricing override from the feedback loop.
	ModelPricingOverride = billingv1.ModelPricingOverride
	// LedgerEntryType filters ledger queries to specific entry types.
	LedgerEntryType = billingv1.LedgerEntryType
	// LedgerView selects a server-resolved slice of the ledger.
	LedgerView = billingv1.LedgerView
)

// AdjustCreditsParams configures a manual credit adjustment.
type AdjustCreditsParams struct {
	OrgID string
	// AmountMicros is positive to add credits, negative to remove.
	AmountMicros int64
	// Reason is recorded on the ledger entry (audit trail).
	Reason string
	// IdempotencyKey deduplicates retries of the same adjustment.
	IdempotencyKey string
}

// GetCreditLedgerParams configures a credit ledger query.
type GetCreditLedgerParams struct {
	OrgID string
	// Page selects a result page; nil returns the server default page.
	Page *Page
	// TypeFilter narrows to specific entry types. Empty means all types.
	TypeFilter []LedgerEntryType
	// StartTime filters to entries on or after this timestamp. Zero means unbounded.
	StartTime time.Time
	// EndTime filters to entries on or before this timestamp. Zero means unbounded.
	EndTime time.Time
	// View selects a server-resolved ledger slice. LedgerView_ledger_view_statement
	// returns only customer-facing money-movement entries and excludes internal
	// mechanics (per-call usage debits, reservation holds/releases). When both
	// View and TypeFilter are set, the effective filter is their intersection.
	View LedgerView
}

// GetBillingUsageReportParams configures a billing usage report query.
type GetBillingUsageReportParams struct {
	OrgID     string
	StartTime time.Time
	EndTime   time.Time
}

// CreateCheckoutSessionParams configures a Stripe Checkout Session.
type CreateCheckoutSessionParams struct {
	OrgID      string
	PackID     string
	SuccessURL string
	CancelURL  string
}

// CreateBillingPortalSessionParams configures a Stripe Billing Portal session.
type CreateBillingPortalSessionParams struct {
	OrgID     string
	ReturnURL string
}

// SetAutoRechargeConfigParams configures automatic credit recharge.
type SetAutoRechargeConfigParams struct {
	OrgID                string
	Enabled              bool
	ThresholdMicros      int64
	RechargeAmountMicros int64
	MonthlyCapMicros     int64
}

// GetCustomerModelPricingParams configures a customer model pricing query.
type GetCustomerModelPricingParams struct {
	// OrgID resolves org-specific policy overrides. Empty for default pricing.
	OrgID string
}

// DecideModelPricingOverrideParams records a decision on a pending pricing override.
type DecideModelPricingOverrideParams struct {
	// OverrideID identifies the PENDING_SIGNOFF override to decide.
	OverrideID string
	// Approve true makes the override ACTIVE (superseding any current ACTIVE
	// override on the same pricing key); false rejects it.
	Approve bool
	// DecisionNote is an optional note recorded on the decision (audit trail).
	DecisionNote string
}

// ListModelPricingBaselinesParams configures a baseline catalog query.
type ListModelPricingBaselinesParams struct {
	// IncludeHistory true includes SUPERSEDED and RETIRED revisions (the full
	// audit history). Default: ACTIVE documents only.
	IncludeHistory bool
}

// UpsertModelPricingBaselineParams creates or revises a baseline entry.
type UpsertModelPricingBaselineParams struct {
	// Baseline is the entry to create or revise, keyed by
	// (modelId, provider, harness). Lifecycle fields (baselineId, status,
	// decision stamps, pricing effectiveAt) are server-owned and ignored.
	Baseline *ModelPricingBaseline
	// RevisionNote is an optional operator note recorded on the revision.
	RevisionNote string
}

// RetireModelPricingBaselineParams retires a model from the registry catalog.
type RetireModelPricingBaselineParams struct {
	ModelID  string
	Provider string
	Harness  string
	// RevisionNote is an optional operator note recorded on the retirement.
	RevisionNote string
}

// BillingClient provides the billing bounded context: account provisioning,
// balance queries, ledger history, credit purchases via Stripe Checkout, and
// the operator model-pricing surfaces.
//
// Internal execution-billing RPCs (authorizeExecution, recordLlmCallUsage,
// finalizeExecution) are not exposed — they are called only by the Temporal
// workflow and the LLM proxy.
type BillingClient struct {
	command billingv1.BillingCommandControllerClient
	query   billingv1.BillingQueryControllerClient
}

func newBillingClient(conn grpc.ClientConnInterface) *BillingClient {
	return &BillingClient{
		command: billingv1.NewBillingCommandControllerClient(conn),
		query:   billingv1.NewBillingQueryControllerClient(conn),
	}
}

// GetOrCreateBillingAccount provisions or retrieves the billing account for
// an organization. Idempotent: creates the account on first call, returns the
// existing account on subsequent calls.
func (b *BillingClient) GetOrCreateBillingAccount(ctx context.Context, orgID string) (*BillingAccount, error) {
	resp, err := b.command.GetOrCreateBillingAccount(ctx, &billingv1.GetOrCreateBillingAccountInput{OrgId: orgID})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// GetBillingAccount retrieves the billing account for an organization.
func (b *BillingClient) GetBillingAccount(ctx context.Context, orgID string) (*BillingAccount, error) {
	resp, err := b.query.GetBillingAccount(ctx, &billingv1.GetBillingAccountInput{OrgId: orgID})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// GetCreditBalance retrieves the credit balance breakdown for an organization.
func (b *BillingClient) GetCreditBalance(ctx context.Context, orgID string) (*CreditBalance, error) {
	resp, err := b.query.GetCreditBalance(ctx, &billingv1.GetCreditBalanceInput{OrgId: orgID})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// AdjustCredits manually adjusts an organization's credit balance.
//
// Positive AmountMicros adds credits (e.g. funding a tenant org), negative
// removes them. The adjustment is recorded as a ledger entry with the
// supplied reason; the idempotency key deduplicates retries. Requires
// can_manage_billing on the org.
func (b *BillingClient) AdjustCredits(ctx context.Context, params *AdjustCreditsParams) (*CreditLedgerEntry, error) {
	resp, err := b.command.AdjustCredits(ctx, &billingv1.AdjustCreditsInput{
		OrgId:          params.OrgID,
		AmountMicros:   params.AmountMicros,
		Reason:         params.Reason,
		IdempotencyKey: params.IdempotencyKey,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// GetCreditLedger retrieves paginated credit ledger entries with optional filters.
func (b *BillingClient) GetCreditLedger(ctx context.Context, params *GetCreditLedgerParams) (*CreditLedgerResponse, error) {
	req := &billingv1.GetCreditLedgerInput{
		OrgId:      params.OrgID,
		TypeFilter: params.TypeFilter,
		View:       params.View,
	}
	if params.Page != nil {
		req.Page = &rpc.PageInfo{Num: params.Page.Num, Size: params.Page.Size}
	}
	if !params.StartTime.IsZero() {
		req.StartTime = timestamppb.New(params.StartTime)
	}
	if !params.EndTime.IsZero() {
		req.EndTime = timestamppb.New(params.EndTime)
	}
	resp, err := b.query.GetCreditLedger(ctx, req)
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// GetBillingUsageReport retrieves an aggregated billing usage report for a
// date range: total provider cost, total billable amount, execution and LLM
// call counts, and a per-model breakdown with cost tier attribution.
func (b *BillingClient) GetBillingUsageReport(ctx context.Context, params *GetBillingUsageReportParams) (*BillingUsageReportResponse, error) {
	resp, err := b.query.GetBillingUsageReport(ctx, &billingv1.GetBillingUsageReportInput{
		OrgId:     params.OrgID,
		StartTime: timestamppb.New(params.StartTime),
		EndTime:   timestamppb.New(params.EndTime),
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// CreateCreditCheckoutSession creates a Stripe Checkout Session to purchase a
// credit pack. The caller should redirect the user to the returned checkout
// URL; credits are provisioned asynchronously via webhook after payment.
func (b *BillingClient) CreateCreditCheckoutSession(ctx context.Context, params *CreateCheckoutSessionParams) (*CreateCreditCheckoutSessionResponse, error) {
	resp, err := b.command.CreateCreditCheckoutSession(ctx, &billingv1.CreateCreditCheckoutSessionInput{
		OrgId:      params.OrgID,
		PackId:     params.PackID,
		SuccessUrl: params.SuccessURL,
		CancelUrl:  params.CancelURL,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// CreateBillingPortalSession creates a Stripe Billing Portal session for
// payment method management. The caller should redirect the user to the
// returned portal URL; changes are synced back via webhooks.
func (b *BillingClient) CreateBillingPortalSession(ctx context.Context, params *CreateBillingPortalSessionParams) (*CreateBillingPortalSessionResponse, error) {
	resp, err := b.command.CreateBillingPortalSession(ctx, &billingv1.CreateBillingPortalSessionInput{
		OrgId:     params.OrgID,
		ReturnUrl: params.ReturnURL,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// SetAutoRechargeConfig configures automatic credit recharge for an
// organization. When enabled, the billing system charges the account's saved
// payment method whenever the available balance drops below the configured
// threshold. Returns the updated BillingAccount.
func (b *BillingClient) SetAutoRechargeConfig(ctx context.Context, params *SetAutoRechargeConfigParams) (*BillingAccount, error) {
	resp, err := b.command.SetAutoRechargeConfig(ctx, &billingv1.SetAutoRechargeConfigInput{
		OrgId:                params.OrgID,
		Enabled:              params.Enabled,
		ThresholdMicros:      params.ThresholdMicros,
		RechargeAmountMicros: params.RechargeAmountMicros,
		MonthlyCapMicros:     params.MonthlyCapMicros,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// GetCustomerModelPricing retrieves the customer-facing model price list with
// markup applied — the prices the customer pays, organized by harness and
// cost tier. Pass params to resolve org-specific policy overrides; nil params
// returns default pricing.
func (b *BillingClient) GetCustomerModelPricing(ctx context.Context, params *GetCustomerModelPricingParams) (*CustomerModelPricingResponse, error) {
	req := &billingv1.GetCustomerModelPricingInput{}
	if params != nil {
		req.OrgId = params.OrgID
	}
	resp, err := b.query.GetCustomerModelPricing(ctx, req)
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// GetModelPricingGovernance retrieves the platform pricing governance view:
// baseline vs effective rates per model, ACTIVE override provenance, and
// pending sign-off proposals from the pricing feedback loop.
//
// Platform-operator surface (can_manage_model_pricing on platform:stigmer):
// rates are raw provider prices, pre-markup.
func (b *BillingClient) GetModelPricingGovernance(ctx context.Context) (*ModelPricingGovernanceResponse, error) {
	resp, err := b.query.GetModelPricingGovernance(ctx, &billingv1.GetModelPricingGovernanceInput{})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// ListModelPricingBaselines retrieves the model registry baseline catalog:
// ACTIVE entries by default, or the full append-only revision history with
// IncludeHistory. Nil params lists ACTIVE entries.
//
// Platform-operator surface (can_manage_model_pricing on platform:stigmer):
// rates are raw provider prices, pre-markup.
func (b *BillingClient) ListModelPricingBaselines(ctx context.Context, params *ListModelPricingBaselinesParams) (*ModelPricingBaselinesResponse, error) {
	req := &billingv1.ListModelPricingBaselinesInput{}
	if params != nil {
		req.IncludeHistory = params.IncludeHistory
	}
	resp, err := b.query.ListModelPricingBaselines(ctx, req)
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// DecideModelPricingOverride records a human decision on a PENDING_SIGNOFF
// pricing override. Approving makes the override ACTIVE (superseding any
// current ACTIVE override on the same pricing key) and recomposes the
// effective registry; rejecting archives it for audit. Returns the decided
// override with the decision stamped.
func (b *BillingClient) DecideModelPricingOverride(ctx context.Context, params *DecideModelPricingOverrideParams) (*ModelPricingOverride, error) {
	resp, err := b.command.DecideModelPricingOverride(ctx, &billingv1.DecideModelPricingOverrideInput{
		OverrideId:   params.OverrideID,
		Approve:      params.Approve,
		DecisionNote: params.DecisionNote,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// UpsertModelPricingBaseline creates or revises one model registry baseline
// entry (catalog + list prices). Append-only: an existing ACTIVE entry for
// the same (modelId, provider, harness) key is superseded, never mutated, and
// the effective registry recomposes immediately. Returns the new revision
// with server-stamped lifecycle fields.
func (b *BillingClient) UpsertModelPricingBaseline(ctx context.Context, params *UpsertModelPricingBaselineParams) (*ModelPricingBaseline, error) {
	resp, err := b.command.UpsertModelPricingBaseline(ctx, &billingv1.UpsertModelPricingBaselineInput{
		Baseline:     params.Baseline,
		RevisionNote: params.RevisionNote,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}

// RetireModelPricingBaseline retires one model from the registry catalog.
// The model disappears from every price surface on the next composition pass;
// the document is kept for audit and the key can be revived by a subsequent
// upsert.
func (b *BillingClient) RetireModelPricingBaseline(ctx context.Context, params *RetireModelPricingBaselineParams) (*ModelPricingBaseline, error) {
	resp, err := b.command.RetireModelPricingBaseline(ctx, &billingv1.RetireModelPricingBaselineInput{
		ModelId:      params.ModelID,
		Provider:     params.Provider,
		Harness:      params.Harness,
		RevisionNote: params.RevisionNote,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return resp, nil
}
